projects/exceptions.py

```diff
@@ -1,30 +1,83 @@
-        self.code = 400
+class ApiException(Exception):
+    """
+    The base exception class for all APIExceptions.
 
+    Parameters
+    ----------
+    code : str
+        Error code.
+    message : str
+        Human readable string describing the exception.
+    status_code : int
+        HTTP status code.
+    """
 
-class Forbidden(Exception):
-    def __init__(self, message: str):
+    def __init__(self, code: str, message: str, status_code: int):
+        self.code = code
         self.message = message
-        self.code = 403
+        self.status_code = status_code
 
 
-class NotFound(Exception):
-    def __init__(self, message: str):
-        self.message = message
-        self.code = 404
+class BadRequest(ApiException):
+    """
+    Bad Request response status code indicates that the server cannot or will not
+    process the request due to something that is perceived to be a client error.
 
+    A common cause is that the client has sent invalid request values.
+    """
 
-class InternalServerError(Exception):
-    def __init__(self, message: str):
-        self.message = message
-        self.code = 500
+    def __init__(self, code: str, message: str):
+        super().__init__(code, message, status_code=400)
 
-class ServiceUnavailable(Exception):
-    def __init__(self, message: str):
-        self.message = message
-        self.code = 503
\ No newline at end of file
+
+class Forbidden(ApiException):
+    """
+    Forbidden client error status response code indicates that the server understands
+    the request but refuses to authorize it.
+    """
+
+    def __init__(self, code: str, message: str):
+        super().__init__(code, message, status_code=403)
+
+
+class NotFound(ApiException):
+    """
+    Not Found client error response code indicates that the server can't find the requested resource.
+
+    A common cause is that a provided ID does not exist in the database.
+    """
+
+    def __init__(self, code: str, message: str):
+        super().__init__(code, message, status_code=404)
+
+
+class InternalServerError(ApiException):
+    """
+    Internal Server Error server error response code indicates that the server
+    encountered an unexpected condition that prevented it from fulfilling the request.
+
+    This error response is a generic "catch-all" response.
+    You should log error responses like the 500 status code with more details about
+    the request to prevent the error from happening again in the future.
+    """
+
+    def __init__(self, code: str, message: str):
+        super().__init__(code, message, status_code=500)
+
+
+class ServiceUnavailable(ApiException):
+    """
+    Service Unavailable server error response code indicates that the server is
+    not ready to handle the request.
+
+    A common cause is that the database is not available or overloaded.
+    """
+
+    def __init__(self, code: str, message: str):
+        super().__init__(code, message, status_code=503)
```

Comment:
@andreluiz27 @dfvneto @dnlcesilva 
Hoje existem estas 5 classes de Exception: `BadRequest`, `Forbidden`, `NotFound`, `InternalServerError` e `ServiceUnavailable`.

Em cada controller as pessoas importam essas classes e criam uma instância delas passando `code=` e `message=`.
Às vezes, há erros que ocorrem várias vezes (por ex: `ProjectNotFound`). Neste caso o código é duplicado em várias áreas do backend, o que é ruim...

Uma alternativa para ajudar com isto é criar mais classes de exception, por exemplo:
```python
class ProjectNotFound(NotFound):
    """
    Indicates that a given project_id does not exist in the database.
    """

    def __init__(self, code: str, message: str):
        code = "ProjectNotFound"
        message = "The specified project does not exist."
        super().__init__(code, message, status_code=404)

```

O que vocês acham desta alternativa? 
Prós: menos código duplicado.
Contra: existiriam umas 50 classes de exception diferentes.

Reply:
Não seria possível utilizar as próprias exceções nativas?

Reply:
@dfvneto as exceções nativas que você disse é esta `HTTPException`?
https://fastapi.tiangolo.com/tutorial/handling-errors/#raise-an-httpexception-in-your-code

Ou alguma outra?

Reply:
Tem ela, e tem o módulo exceptions do fastapi também. Não tenho certeza do fastapi, mas no django-rest-framework e no flask quando eu trabalhei dava para utilizar os módulos de exceções tranquilamente, menos quando tinha algum problema muito específico.

Reply:
![image](https://user-images.githubusercontent.com/30246766/142436600-a187e945-886c-4768-885a-475a40c573f1.png)


....

Mas não tem quase nada implementado no módulo de exceções do fast api. Divertido